File: generators/generator.m
% generate data for clustering
% we will generate 3 clusters
% cluster 1: number of samples n1
  n1 = 42;
% taxpayer age -- normal distribution with mean 30 and variance 5
  tx = normrnd(30,5,1,n1);
% walkability -- normal distribution with mean 90 and variance of 5
  w = normrnd(90,5,1,n1);
% distance to park -- normal distribution with mean of 2 and variance of
% 0.3
  dp = normrnd(2,0.3,1,n1);
% square footage -- normal distribution with mean of 1400 and variance of
% 100
  sqf = normrnd(1400,100,1,n1);
% lot size -- normal distribution with mean of 0.25 and variance of 0.05
  lot = normrnd(0.25,0.05,1,n1);
% bedrooms -- integer random {2,3,4} average around 3 with ratio 20% are
% 2's
  a = inv([ 3 4 ; 1 1])*[3*n1 - 0.2*2*n1; n1 - 0.2*n1];
  a1 = floor(0.2*n1); a2 = floor(a(1)); a3 = n1 - a1 - a2;
  temp = [2*ones(1,a1) 3*ones(1,a2) 4*ones(1,a3)];
  br = temp(randperm(n1));
% bathrooms -- integer random {1,2} average around 1.8
  a2 = floor(0.8*n1); a1 = n1 - a2;
  temp = [1*ones(1,a1) 2*ones(1,a2)];
  bh = temp(randperm(n1));
% garage -- integer random {1,2} average around 1.7
  a2 = floor(0.7*n1); a1 = n1 - a2;
  temp = [1*ones(1,a1) 2*ones(1,a2)];
  g = temp(randperm(n1));
% off street parking -- integer random {1,2} average around 1.6
  a2 = floor(0.6*n1); a1 = n1 - a2;
  temp = [1*ones(1,a1) 2*ones(1,a2)];
  os = temp(randperm(n1));
% distance to public schools -- integer {0,1} average around 0.8
  a2 = floor(0.8*n1); a1 = n1 - a2;
  temp = [zeros(1,a1) 1*ones(1,a2)];
  pubsc = temp(randperm(n1));
% distance to private schools -- integer {0,1} average around 0.2
  a2 = floor(0.2*n1); a1 = n1 - a2;
  temp = [zeros(1,a1) 1*ones(1,a2)];
  prvsc = temp(randperm(n1));
% access to highway -- integer {0,1} average 0.7
  a2 = floor(0.7*n1); a1 = n1 - a2;
  temp = [zeros(1,a1) 1*ones(1,a2)];
  hway = temp(randperm(n1));
% export as xls file
  A = {'tax age'; 'walk sc'; 'dist to park'; 'sq ft'; 'lot sz'; 'bedroom'; ...
       'bathroom'; 'garage'; 'off street'; 'pub school'; 'prv school'; 'hway acces'};
  B = [ tx; w; dp; sqf; lot; br; bh; g; os; pubsc; prvsc; hway];

  xlswrite('junk.xls',A);
  xlswrite('junk.xls',B,1,'B1');

File: generators/generator.py
#!/usr/bin/env python

import math
import random
import numpy as np
from scipy.stats import norm
from scipy.linalg import inv

# generate data for clustering
# we will generate 3 clusters
# cluster 1: number of samples NUM_SAMPLES
NUM_SAMPLES = 42;

# taxpayer age -- normal distribution with mean 35 and variance 5
taxpayerAge = np.round(norm.rvs(
    loc = 35,
    scale = 5,
    size = NUM_SAMPLES
), decimals = 1)

# walkability -- normal distribution with mean 90 and variance of 5
walkability = np.round(norm.rvs(
    loc = 90,
    scale = 5,
    size = NUM_SAMPLES
), decimals = 1)

# distance to park -- normal distribution with mean of 2 and variance of 0.3
distanceToPark = np.round(norm.rvs(
    loc = 2,
    scale = 0.3,
    size = NUM_SAMPLES
), decimals = 1)

# square footage -- normal distribution with mean of 1400 and variance of 500
squareFootage = np.round(norm.rvs(
    loc = 1400,
    scale = 500,
    size = NUM_SAMPLES
))

# lotSize size -- normal distribution with mean of 0.25 and variance of 0.05
lotSize = np.round(norm.rvs(
    loc = 0.25,
    scale = 0.05,
    size = NUM_SAMPLES
), decimals = 2)

# function that naively generates a set with the given size
# using the given source numbers and with approximately the given mean
def generateSamples(source = [], loc = 0, size = NUM_SAMPLES):
    source.sort()
    source = np.array(source)

    samples = np.empty(size)
    samples[0] = random.choice(source)

    i = 1
    while i < size:
        mean = np.mean(samples[:i])
        if mean < loc:
            nextOpts = np.extract(mean < source, source)
        elif loc > mean:
            nextOpts = np.extract(source > mean, source)
        else:
            nextOpts = source
        samples[i] = random.choice(nextOpts)
        i += 1

    random.shuffle(samples)
    return samples

# bedrooms -- integer random {2, 3, 4} average around 3
bedrooms = generateSamples(
    source = [2, 3, 4],
    loc = 3
)

# bathrooms -- float random {1.0, 1.5, 2.0} average around 1.8
bathrooms = generateSamples(
    source = [1, 1.5, 2],
    loc = 1.8
)

# garage -- integer random {0, 1, 2} average around 1.7
garage = generateSamples(
    source = [0, 1, 2],
    loc = 1.7
)

# off street parking -- integer random {0, 1, 2} average around 1.6
parking = generateSamples(
    source = [0, 1, 2],
    loc = 1.6
)

# within walking distance of public schools -- integer {0, 1} average around 0.4
publicSchools = generateSamples(
    source = [0, 1],
    loc = 0.4
)

# within walking distance of private schools -- integer {0, 1} average around 0.2
privateSchools = generateSamples(
    source = [0, 1],
    loc = 0.2
)

# access to highway -- integer {0, 1} average 0.7
highway = generateSamples(
    source = [0, 1],
    loc = 0.4
)

# export as csv file
# human readable headers
header = [
    'Age',
    'Walkability',
    'Distance to Park',
    'Square Footage',
    'Lot Acreage',
    'Bedrooms',
    'Bathrooms',
    'Garage Size',
    'Off-Street Parking',
    'Walkable Public School',
    'Walkable Private School',
    'Highway Access'
]

# ensure all generated data are positive
# transpose data so that each row is a house instead of each column
data = np.absolute(np.transpose([
    taxpayerAge,
    walkability,
    distanceToPark,
    squareFootage,
    lotSize,
    bedrooms,
    bathrooms,
    garage,
    parking,
    publicSchools,
    privateSchools,
    highway
]))

np.savetxt('junk.tsv', data,
    header = '\t'.join(header), # Add column names as comment instead of row
    fmt = '%s',                 # Format all cells as strings
    delimiter = '\t'            # tsv
)
